types/types.ts

```diff
@@ -4,22 +4,25 @@ export interface WerkitMeta {
   runtime_info?: any
 }
 
-export interface WerkitSuccessResult<ResultType> extends WerkitMeta {
+export interface WerkitSuccessResult<ResultType, MessageKeyType>
+  extends WerkitMeta {
   success: true
   result: ResultType
   error: null
   error_origin: null
+  message_key: MessageKeyType
 }
 
 export type WerkitErrorOrigin = 'compute' | 'system' | 'orchestration'
 
-export interface WerkitErrorResult extends WerkitMeta {
+export interface WerkitErrorResult<MessageKeyType> extends WerkitMeta {
   success: false
   result: null
-  error: string[]
+  error: string[] | Error
```

Comment:
This needs to be something JSON-serializable and cross platform. In Python we're serializing the stack trace as an array of newline-terminated strings. I guess I'd suggest we do something similar on the Node side.

Reply:
> This needs to be something JSON-serializable and cross platform. 

Fair enough. Error itself is not JSON-serializable. However, I do like the properties it exposes, which are message, name and stack (you can see this [here](https://github.com/curvewise/vogue/pull/46/files#diff-9b1c70fddbc960215ec54e116f57f4af1a5deb9d573fc6c4211f6afbd07c0f44R145-R147)). So, two follow-up questions:

How do you feel about passing through an Error-like object with these properties?
If you are alright with letting the `error` property be this Error-like object, would you prefer we define our own typescript type, or keep the Error type?

Reply:
The goal of this type is to document a JSON schema. We're using TypeScript to generate it because it's much friendlier to read, though this type should be a subset of JSON.

It would be fine to create a utility function to accept a Node Error and turn it into a werkit-serializable error (which could even go in this package). However this type definition should generate a JSON Schema which can be used to validate whether the message is correct.

In Python the name and message are already part of the stack trace:

```
>>> os.makedirs('/foobar')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/foobar'
```

This behavior is under test:

https://github.com/metabolize/werkit/blob/d1c72bcbd3bb499f1f352a8d765ca36af919973d/werkit/test_manager.py#L41

We don't yet have a need to do any analytics on these errors, and it is less work to return this stack trace than to model the error, so I think the simplest thing to do would be to match the Node error to Python's way of modeling this, which is to basically serialize that last line if it's not already part of the stacktrace.

If you don't want to do that, I think we need to figure out how to either unify the data structure across the two runtimes, so we can easily present something consistent-looking in the UI, or else go for the really thorough thing which would be to have the schema be a little different depending on the runtime. So `error` could be a union type with a sub-key `runtime: 'python'` or `runtime: 'node'`, and the frontend can adjust how it displays it.

There are a few options here, and while I don't want to push all of the complexity to the consumer of the errors to figure out how to format these keys depending on who sent them, which feels developer-unfriendly in the long term, I'm good with the other approaches, of either all-out modeling different error schema, or harmonizing our Node errors with the way Python formats the stack traces.

Reply:
*tl;dr:* Right now, I think it would be best for node to adopt the existing werkit model of sending the error as a stack trace with a string[], but I think we should revisit this decision later, if and when we encounter use cases for which this model is not the best fit.

Here's how I arrived at that conclusion:

This is how node formats the stack track: 

```
function foo(){
  throw new Error('bar')
}

try { 
  foo()
} catch (e) {
  console.log(e.stack)
}
```


```
Error: bar
    at foo (/home/jacob/vimwiki/test.js:2:9)
    at Object.<anonymous> (/home/jacob/vimwiki/test.js:6:3)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)
```

So, I could move the first line of the stack trace to the last line, and it should match the python representation pretty well.

It's useful to note that we *could* extract the same name and message properties from the Python error, like this:

```
[jacob@localhost vimwiki]$ python
Python 3.7.4 (default, Jul  9 2019, 16:48:28) 
[GCC 8.3.1 20190223 (Red Hat 8.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> e = Exception('bar')
>>> e.args
('bar',)
>>> type(e).__name__
'Exception'
>>> 
```

So, we could also create unified model based on the JavaScript error model, if we want to. I think this means that there is no need for a runtime-dependent model - a unified model based on either storing the stack trace in a string[], or based on modeling the error, should work for both runtimes. So, the question then becomes, what are the advantages and disadvantages of sending the modeled error, versus just sending the stack trace? 

It seems to me that while the stack trace captures both the name (Error) and message (bar) properties of the error, the disadvantage that I see with this approach is that, if the consumer wants to read these properties, it will need to parse them out, as you're doing with the werkit test manager above. For example, if it needs to handle different types of errors in different ways, then it will need to parse the stack trace to find this information. So, the question is, will we have a use case where we *will* want to parse this information? And I think the answer to that is also yes. Even for the purpose of displaying the error message to the client in the UI, you would probably want to show just the error name and message, then probably dump the full stack trace to console.error for a developer to read. 

On the other hand, the advantage to the approach of just sending an error string (or string array), is that we already have code written around this, so it is more compatible with the existing approach, and there is less code to change. Given that we want to wrap this feature up ASAP, I think we want to minimize the amount of time we spend refactoring existing code, and so that means that we should give maintaining compatibility with the existing model a lot of weight at this moment. Therefore, I think that while modeling the error is better in a lot of ways, and will probably be a better fit for our requirements in the long term, I think in the short term, we should choose to maintain compatibility with the existing model. If you are OK with that, I will go ahead and make these changes to the vogue lambda and the goldie-data-layer vogue task handler.

Reply:
Fixed in 00ad576b2eff6c2c0ad413bc08d2fa39ec1beeb2 and curvewise/vogue@0f1fa4691a3a9d762ed4439b17f86319d18a8566

Reply:
Wholeheartedly agree with all of this.

Since the Python and Node formatted stack traces are already pretty similar, I do not think we need to reassemble them to make them _even more_ similar. In other words, I think it is fine to show them in their native way for now: Node, with error/message on top, and Python, with error/message on the bottom.

Makes sense that we'd approach modeling the errors in a consistent way when we might need to.

In the case of the compute code, a user may need to know that there was an error, however I do not think the error type and error message will be helpful to the developer. The code I'm thinking of isn't intended to ever fail, and the error messages are typically far too cryptic for a non-developer to debug. So the audience for the error and the stack trace may be the same. Though if that turns out to be wrong, and we do want to show the error / message separate from the stack trace, I completely agree that we should revisit the serialization format at that time.